### PR TITLE
ML-034/improve role system

### PR DESCRIPTION
**Pull request by WhatWasThatt**
# Description

add new permissions
delete excess permissions
update policies

## Trello card link

https://trello.com/c/zSuiwEfF

# Checklist:

- [x] My code fixes the problem or adds required functionality
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes generate no new warnings
- [ ] I have added tests that prove that my feature works
- [x] New and existing unit tests pass locally with my changes

# Additional requirements:

- [x] Does your code require any migrations to be ran?
- [ ] Does your code require installing any new gems or system dependencies?
- [ ] Does your code require any changes to current deployment process?

# Additional Information

### CODE REVIEW

File: db/seeds.rb [hunk 1]
```diff
@@ -8,13 +8,16 @@
...
+                          can_read_entities: true, can_edit_entities: true })
 redactor_role = Role.create(name: 'redactor', permission_attributes:
                         { can_lock_user: false, can_edit_role: false,
-                          can_set_role: false, can_read: true, can_edit: true })
+                          can_set_role: false, can_edit_movie_and_person: true, can_read_movie_and_person: true,
+                          can_read_entities: true, can_edit_entities: true })
 reviewer_role = Role.create(name: 'reviewer', permission_attributes:
                         { can_lock_user: false, can_edit_role: false,
-                          can_set_role: false, can_read: true, can_edit: false })
+                          can_set_role: false, can_edit_movie_and_person: true, can_read_movie_and_person: true,
```
Comment (twill11) [inline db/seeds.rb:19]: do reviewer actually 'can_edit_movie_and_person' can do this?

File: app/policies/access_user_policy.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+    def resolve
+      scope.all
+    end
+
+    private
+
+    attr_reader :user, :scope
+  end
+end
```
Comment (twill11) [inline app/policies/access_user_policy.rb:22]: are you sure we need all these scopes?

File: app/models/role.rb [hunk 3]
```diff
@@ -49,17 +57,17 @@ def reviewer?
...
 
   def editor?
     [admin?, redactor?].include?(name)
   end
+
+  private
+
+  def permision?
+    !permission.nil?
+  end
```
Comment (dmitrymalllpr) [inline app/models/role.rb:72]: It s very RUDE. Please use git more better

**Review by dmitrymalllpr — CHANGES_REQUESTED** [review]

File: app/policies/access_user_policy.rb (hunk 2 above)
Comment (WhatWasThatt) [reply app/policies/access_user_policy.rb]: yes, you are right

File: app/models/role.rb (hunk 3 above)
Comment (WhatWasThatt) [reply app/models/role.rb:72]: It means that predefined_user and editor methods become not private

File: app/models/role.rb [hunk 4]
```diff
@@ -33,8 +29,20 @@ def can_set_role?
...
+  def can_edit_entities?
+    editor? ? true : permision? && permission.can_edit_entities
+  end
+
+  def can_read_movie_and_person?
+    predefined_user? ? true : permision? && permission.can_read_movie_and_person
+  end
+
+  def can_edit_movie_and_person?
+    editor? ? true : permision? && permission.can_edit_movie_and_person
```
Comment (dmitrymalllpr) [inline app/models/role.rb:45]: solid and kiss. If you don t see problem please write me

File: app/models/permission.rb [hunk 5]
```diff
@@ -2,15 +2,17 @@
...
+#  can_edit_entities         :boolean
+#  can_edit_movie_and_person :boolean
+#  can_edit_role             :boolean
+#  can_lock_user             :boolean
+#  can_read_entities         :boolean
+#  can_read_movie_and_person :boolean
+#  can_set_role              :boolean
+#  created_at                :datetime         not null
+#  updated_at                :datetime         not null
+#  role_id                   :bigint           not null
```
Comment (dmitrymalllpr) [inline app/models/permission.rb:15]: looks strange. Please try use git better

File: app/policies/permission_policy.rb [hunk 6]
```diff
@@ -1,17 +1,8 @@
...
-    super(record)
-  end
-
-  def index?
-    user.can_read?
-  end
-
   def show?
-    user.can_read?
+    user.can_edit_role?
```
Comment (dmitrymalllpr) [inline app/policies/permission_policy.rb:5]: I am not sure it is good naming

File: app/models/user.rb [hunk 7]
```diff
@@ -59,8 +59,7 @@ class User < ApplicationRecord
 
   belongs_to :role, optional: true
   has_many :ratings, dependent: :destroy
-
```
Comment (dmitrymalllpr) [inline app/models/user.rb:62]: Why you deleted empty line?

File: app/policies/user_policy.rb [hunk 8]
```diff
@@ -1,31 +1,18 @@
...
   attr_reader :user, :record
 
-  def initialize(user, record)
-    super(user)
-    super(record)
-  end
-
   def index?
-    user.can_read?
+    user.predefined_user?
```
Comment (dmitrymalllpr) [inline app/policies/user_policy.rb:5]: Why only .predefined_user? Please think about policy

**Review by dmitrymalllpr — APPROVED** [review]

File: app/controllers/api/v1/permissions_controller.rb [hunk 9]
```diff
@@ -36,7 +36,7 @@ def set_permission
       end
 
       def permission_params
-        params.require(:permission).permit(:can_lock_user, :can_edit_role, :can_set_role, :can_read, :can_edit)
+        params.require(:permission).permit(:can_lock_user, :can_edit_role, :can_set_role, :can_edit_movie_and_person, :can_read_movie_and_person, :can_read_entities, :can_edit_entities)
```
Comment (dmitrymalllpr) [inline app/controllers/api/v1/permissions_controller.rb:39]: Maybe line is too long

**Review by dmitrymalllpr — CHANGES_REQUESTED** [review]

File: app/policies/country_policy.rb [hunk 10]
```diff
@@ -1,2 +1,23 @@
-class CountryPolicy < BasePolicy
```
Comment (dmitrymalllpr) [inline app/policies/country_policy.rb:1]: Why you removed BasePolicy.  What wrong with this line?

**Review by dmitrymalllpr — CHANGES_REQUESTED** [review]

File: app/models/permission.rb (hunk 5 above)
Comment (WhatWasThatt) [reply app/models/permission.rb:15]: but it annotates comments, it aligns them, not git

File: app/models/user.rb (hunk 7 above)
Comment (WhatWasThatt) [reply app/models/user.rb:62]: it's prettier

Comment (dmitrymalllpr) [reply app/models/user.rb:62]: I don t ask you revert it. But please try create more accurate commits!

File: app/models/permission.rb (hunk 5 above)
Comment (dmitrymalllpr) [reply app/models/permission.rb]: exactly! But it is NOT the best practice. YOU should write code  satisfied to linters requirements

File: app/models/role.rb (hunk 4 above)
Comment (dmitrymalllpr) [reply app/models/role.rb:53]: You don t break first solid  principe. IT about classes not methods. BUT methods MUST have one responsibility TOO. I think you DON T follow KISS. You method too complicated and NOT universal. PLEASE can_edit_movie? can_edit_smt? BUT avoid AND

Comment (WhatWasThatt) [reply app/models/role.rb:53]: ok thanks, i'll fix it now

**Comment by dmitrymalllpr** [comment]
Maybe make sense to follow more strict rubocops rules?

**Review by dmitrymalllpr — COMMENTED** [review]
38 files!!!!

**Review by dmitrymalllpr — APPROVED** [review]

**Review by twill11 — APPROVED** [review]
GJ!

**Pull request merged**